### ISSUE TITLE
[NixOS] Failed to Build

### ISSUE THREAD

**Issue by PixeKnight** [horriblename/hyprgrass#187]
**Describe the bug**
Ninja failed to build package

**Log**
output of `nix log /nix/store/bajflgyxq2nmfsyqrkn8lqzs60q4fqx9-hyprgrass-v0.8.2-.drv`

```
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/bajflgyxq2nmfsyqrkn8lqzs60q4fqx9-hyprgrass-v0.8.2-.drv^*'
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/f777pls7mqcbfi3bzii4dmmyd2rgmzyz-9hx74y52klj5kvgyql67dq8f1bax7qky-source
source root is 9hx74y52klj5kvgyql67dq8f1bax7qky-source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
mesonConfigurePhase flags: --prefix=/nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2- --libdir=/nix/store/xxwjd>
The Meson build system
Version: 1.6.0
Source dir: /build/9hx74y52klj5kvgyql67dq8f1bax7qky-source
Build dir: /build/9hx74y52klj5kvgyql67dq8f1bax7qky-source/build
Build type: native build
Project name: hyprgrass
Project version: v0.8.2
C compiler for the host machine: gcc (gcc 13.3.0 "gcc (GCC) 13.3.0")
C linker for the host machine: gcc ld.bfd 2.43.1
C++ compiler for the host machine: g++ (gcc 13.3.0 "g++ (GCC) 13.3.0")
C++ linker for the host machine: g++ ld.bfd 2.43.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C++ supports arguments -std=c++23: YES 
meson.build:8: WARNING: Consider using the built-in option for language standard version instead of using "-std=c++23".
Found pkg-config: YES (/nix/store/16h31npnbgpkhxp1041dsifiklbw4mpr-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Run-time dependency hyprland found: YES 0.45.0
Run-time dependency wftouch found: YES 0.0
Found CMake: /nix/store/4k2h9qsps52n1v56zq03yi67dwmxa81i-cmake-3.30.5/bin/cmake (3.30.5)
Run-time dependency doctest found: YES 2.4.11
Configuring version.hpp using configuration
Run-time dependency pixman-1 found: YES 0.43.4
Run-time dependency libinput found: YES 1.26.2
Run-time dependency wayland-server found: YES 1.23.1
Run-time dependency xkbcommon found: YES 1.7.0
Run-time dependency libdrm found: YES 2.4.123
Build targets in project: 3

hyprgrass v0.8.2

  User defined options
    auto_features: enabled
    bindir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/bin
    buildtype    : plain
    includedir   : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/include
    infodir      : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/info
    libdir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/lib
    libexecdir   : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/libexec
    localedir    : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/locale
    mandir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/man
    prefix       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-
    sbindir      : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/sbin
    wrap_mode    : nodownload

lines 1-55...skipping...
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/f777pls7mqcbfi3bzii4dmmyd2rgmzyz-9hx74y52klj5kvgyql67dq8f1bax7qky-source
source root is 9hx74y52klj5kvgyql67dq8f1bax7qky-source
Running phase: patchPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/f777pls7mqcbfi3bzii4dmmyd2rgmzyz-9hx74y52klj5kvgyql67dq8f1bax7qky-source
source root is 9hx74y52klj5kvgyql67dq8f1bax7qky-source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
mesonConfigurePhase flags: --prefix=/nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2- --libdir=/nix/store/xxwjd>
The Meson build system
Version: 1.6.0
Source dir: /build/9hx74y52klj5kvgyql67dq8f1bax7qky-source
Build dir: /build/9hx74y52klj5kvgyql67dq8f1bax7qky-source/build
Build type: native build
Project name: hyprgrass
Project version: v0.8.2
C compiler for the host machine: gcc (gcc 13.3.0 "gcc (GCC) 13.3.0")
C linker for the host machine: gcc ld.bfd 2.43.1
C++ compiler for the host machine: g++ (gcc 13.3.0 "g++ (GCC) 13.3.0")
C++ linker for the host machine: g++ ld.bfd 2.43.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C++ supports arguments -std=c++23: YES 
meson.build:8: WARNING: Consider using the built-in option for language standard version instead of using "-std=c++23".
Found pkg-config: YES (/nix/store/16h31npnbgpkhxp1041dsifiklbw4mpr-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Run-time dependency hyprland found: YES 0.45.0
Run-time dependency wftouch found: YES 0.0
Found CMake: /nix/store/4k2h9qsps52n1v56zq03yi67dwmxa81i-cmake-3.30.5/bin/cmake (3.30.5)
Run-time dependency doctest found: YES 2.4.11
Configuring version.hpp using configuration
Run-time dependency pixman-1 found: YES 0.43.4
Run-time dependency libinput found: YES 1.26.2
Run-time dependency wayland-server found: YES 1.23.1
Run-time dependency xkbcommon found: YES 1.7.0
Run-time dependency libdrm found: YES 2.4.123
Build targets in project: 3

hyprgrass v0.8.2

  User defined options
    auto_features: enabled
    bindir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/bin
    buildtype    : plain
    includedir   : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/include
    infodir      : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/info
    libdir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/lib
    libexecdir   : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/libexec
    localedir    : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/locale
    mandir       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/share/man
    prefix       : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-
    sbindir      : /nix/store/xxwjd6k0qf4kiz2d6bm4g6cs88hdaz52-hyprgrass-v0.8.2-/sbin
    wrap_mode    : nodownload

Found ninja-1.12.1 at /nix/store/s2lws03pga0vyq84kfa6h041n8csbg4p-ninja-1.12.1/bin/ninja
mesonConfigurePhase: enabled\ parallel\ building
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: -j8
[1/13] Compiling C++ object src/gestures/libgestures.a.p/Shared.cpp.o
[2/13] Compiling C++ object src/gestures/libgestures.a.p/DragGesture.cpp.o
[3/13] Compiling C++ object src/gestures/libgestures.a.p/CompletedGesture.cpp.o
[4/13] Compiling C++ object src/gestures/test/test-gestures.p/MockGestureManager.cpp.o
[5/13] Compiling C++ object src/gestures/libgestures.a.p/Actions.cpp.o
[6/13] Compiling C++ object src/gestures/libgestures.a.p/Gestures.cpp.o
[7/13] Linking static target src/gestures/libgestures.a
[8/13] Compiling C++ object src/libhyprgrass.so.p/VecSet.cpp.o
../src/VecSet.cpp: In instantiation of 'bool VecSet<T>::remove(T) [with T = Hyprutils::Memory::CWeakPointer<CWLTouchResourc>
../src/VecSet.hpp:27:16:   required from here
../src/VecSet.cpp:23:23: warning: comparison of integer expressions of different signedness: 'int' and 'std::vector<Hypruti>
   23 |     for (int i = 0; i < this->set.size(); i++) {
      |                     ~~^~~~~~~~~~~~~~~~~~
../src/VecSet.cpp:25:19: warning: comparison of integer expressions of different signedness: 'int' and 'std::vector<Hypruti>
   25 |             if (i != this->set.size() - 1) {
      |                 ~~^~~~~~~~~~~~~~~~~~~~~~~
[9/13] Compiling C++ object src/libhyprgrass.so.p/main.cpp.o
FAILED: src/libhyprgrass.so.p/main.cpp.o 
g++ -Isrc/libhyprgrass.so.p -Isrc -I../src -I/nix/store/1mvw0dlm8p5mjmqfgmnvnnzb0rpa3bc7-wf-touch-git-dev/include -I/nix/st>
In file included from ../src/main.cpp:1:
../src/GestureManager.hpp:28:22: error: 'ITouch' has not been declared
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |                      ^~~~~~
../src/GestureManager.hpp:28:41: error: expected ',' or '...' before 'e'
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |                                         ^
../src/GestureManager.hpp:32:20: error: 'ITouch' has not been declared
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |                    ^~~~~~
../src/GestureManager.hpp:32:37: error: expected ',' or '...' before 'e'
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |                                     ^
../src/GestureManager.hpp:36:22: error: 'ITouch' has not been declared
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |                      ^~~~~~
../src/GestureManager.hpp:36:43: error: expected ',' or '...' before 'e'
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |                                           ^
../src/main.cpp: In function 'void hkOnTouchDown(void*, SCallbackInfo&, std::any)':
../src/main.cpp:21:55: error: cannot convert 'ITouch::SDownEvent' to 'int'
   21 |     cbinfo.cancelled = g_pGestureManager->onTouchDown(ev);
      |                                                       ^~
      |                                                       |
      |                                                       ITouch::SDownEvent
../src/GestureManager.hpp:28:30: note:   initializing argument 1 of 'bool GestureManager::onTouchDown(int)'
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |                      ~~~~~~~~^~~~~~~~~~
../src/main.cpp: In function 'void hkOnTouchUp(void*, SCallbackInfo&, std::any)':
../src/main.cpp:27:53: error: cannot convert 'ITouch::SUpEvent' to 'int'
   27 |     cbinfo.cancelled = g_pGestureManager->onTouchUp(ev);
      |                                                     ^~
      |                                                     |
      |                                                     ITouch::SUpEvent
../src/GestureManager.hpp:32:28: note:   initializing argument 1 of 'bool GestureManager::onTouchUp(int)'
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |                    ~~~~~~~~^~~~~~~~
../src/main.cpp: In function 'void hkOnTouchMove(void*, SCallbackInfo&, std::any)':
../src/main.cpp:33:55: error: cannot convert 'ITouch::SMotionEvent' to 'int'
   33 |     cbinfo.cancelled = g_pGestureManager->onTouchMove(ev);
      |                                                       ^~
      |                                                       |
      |                                                       ITouch::SMotionEvent
../src/GestureManager.hpp:36:30: note:   initializing argument 1 of 'bool GestureManager::onTouchMove(int)'
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |                      ~~~~~~~~^~~~~~~~~~~~
../src/main.cpp: In function 'void listInternalBinds(std::string)':
../src/main.cpp:41:16: error: 'LogLevel' has not been declared
   41 |     Debug::log(LogLevel::LOG, "[hyprgrass] Listing internal binds:");
      |                ^~~~~~~~
../src/main.cpp:43:20: error: 'LogLevel' has not been declared
   43 |         Debug::log(LogLevel::LOG, "[hyprgrass] | gesture: {}", bind->key);
      |                    ^~~~~~~~
../src/main.cpp:44:20: error: 'LogLevel' has not been declared
   44 |         Debug::log(LogLevel::LOG, "[hyprgrass] |     dispatcher: {}", bind->handler);
      |                    ^~~~~~~~
../src/main.cpp:45:20: error: 'LogLevel' has not been declared
   45 |         Debug::log(LogLevel::LOG, "[hyprgrass] |     arg: {}", bind->arg);
      |                    ^~~~~~~~
../src/main.cpp:46:20: error: 'LogLevel' has not been declared
   46 |         Debug::log(LogLevel::LOG, "[hyprgrass] |     mouse: {}", bind->mouse);
      |                    ^~~~~~~~
../src/main.cpp:47:20: error: 'LogLevel' has not been declared
   47 |         Debug::log(LogLevel::LOG, "[hyprgrass] |");
      |                    ^~~~~~~~
[10/13] Compiling C++ object src/libhyprgrass.so.p/GestureManager.cpp.o
FAILED: src/libhyprgrass.so.p/GestureManager.cpp.o 
g++ -Isrc/libhyprgrass.so.p -Isrc -I../src -I/nix/store/1mvw0dlm8p5mjmqfgmnvnnzb0rpa3bc7-wf-touch-git-dev/include -I/nix/st>
In file included from ../src/GestureManager.cpp:1:
../src/GestureManager.hpp:28:22: error: 'ITouch' has not been declared
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |                      ^~~~~~
../src/GestureManager.hpp:28:41: error: expected ',' or '...' before 'e'
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |                                         ^
../src/GestureManager.hpp:32:20: error: 'ITouch' has not been declared
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |                    ^~~~~~
../src/GestureManager.hpp:32:37: error: expected ',' or '...' before 'e'
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |                                     ^
../src/GestureManager.hpp:36:22: error: 'ITouch' has not been declared
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |                      ^~~~~~
../src/GestureManager.hpp:36:43: error: expected ',' or '...' before 'e'
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |                                           ^
../src/GestureManager.cpp:352:6: error: no declaration matches 'bool GestureManager::onTouchDown(ITouch::SDownEvent)'
  352 | bool GestureManager::onTouchDown(ITouch::SDownEvent ev) {
      |      ^~~~~~~~~~~~~~
../src/GestureManager.hpp:28:10: note: candidate is: 'bool GestureManager::onTouchDown(int)'
   28 |     bool onTouchDown(ITouch::SDownEvent e);
      |          ^~~~~~~~~~~
../src/GestureManager.hpp:19:7: note: 'class GestureManager' defined here
   19 | class GestureManager : public IGestureManager {
      |       ^~~~~~~~~~~~~~
../src/GestureManager.cpp:413:6: error: no declaration matches 'bool GestureManager::onTouchUp(ITouch::SUpEvent)'
  413 | bool GestureManager::onTouchUp(ITouch::SUpEvent ev) {
      |      ^~~~~~~~~~~~~~
../src/GestureManager.hpp:32:10: note: candidate is: 'bool GestureManager::onTouchUp(int)'
   32 |     bool onTouchUp(ITouch::SUpEvent e);
      |          ^~~~~~~~~
../src/GestureManager.hpp:19:7: note: 'class GestureManager' defined here
   19 | class GestureManager : public IGestureManager {
      |       ^~~~~~~~~~~~~~
../src/GestureManager.cpp:465:6: error: no declaration matches 'bool GestureManager::onTouchMove(ITouch::SMotionEvent)'
  465 | bool GestureManager::onTouchMove(ITouch::SMotionEvent ev) {
      |      ^~~~~~~~~~~~~~
../src/GestureManager.hpp:36:10: note: candidate is: 'bool GestureManager::onTouchMove(int)'
   36 |     bool onTouchMove(ITouch::SMotionEvent e);
      |          ^~~~~~~~~~~
../src/GestureManager.hpp:19:7: note: 'class GestureManager' defined here
   19 | class GestureManager : public IGestureManager {
      |       ^~~~~~~~~~~~~~
[11/13] Compiling C++ object src/gestures/test/test-gestures.p/test.cpp.o
ninja: build stopped: subcommand failed.
```

**Desktop (please complete the following information):**
 - OS: [NixOS]
 - output of `hyprctl version` or `Hyprland -v`:
```
Hyprland 0.44.0 built from branch  at commit e58e97b0a38b8ccc87a4304c9e4e2b37c9966875  ().
Date: 2024-11-07
Tag: , commits: 
built against aquamarine 0.4.3


flags set:
debug
```

- output of `hyprctl plugin list`:
```
Plugin borders-plus-plus by Vaxry:
	Handle: 2ed9df30
	Version: 1.0
	Description: A plugin to add more borders to windows.
```

**Additional context**
To recreate, have  

```
wayland.windowManager.hyprland = {
   plugins = [
      inputs.hyprgrass.packages.${pkgs.system}.default
   ];
};
```

in your home-manager module, and

```
{
    inputs = {
       # ...
      hyprland.url = "github:hyprwm/Hyprland";
      hyprgrass = {
         url = "github:horriblename/hyprgrass";
         inputs.hyprland.follows = "hyprland"; # IMPORTANT
      };
  };
}
```

in your nix flake, and run
`nixos-rebuild switch --flake /etc/nixos#default`

Perhaps could also be something to do with the impermenance module I have, but I doubt it, as the issue is during the build, and not reboot.

And just in case, I did include the dependany packages in case that's not included within the nix flake itself.

Also, while it says I have borders++, I removed it before rebuilding with hyprgrass (realised it was static and didn't like it)

**Comment by caffeine01** [comment]
1. Your hyprland version is outdated, update it. (nix flake update).
2. fixed in #186

**Comment by PixeKnight** [comment]
Omg, thanks so much.
